samples/core/callback/main.cpp

```diff
@@ -0,0 +1,341 @@
+        {
+            throw std::runtime_error("Required image format is not supported "
+                                     "on the selected runtime");
+        }
+
+        // Options provided on the command line
+        const std::size_t side = alg_opts.side;
+        const std::size_t iterations = alg_opts.iterations;
+        const std::size_t save_at_every = alg_opts.write_iter;
+
+        // Create two equivalent images. In a single iteration one serves as the
+        // source, the other as the destination, and then the roles are swapped.
+        DoubleBuffer<cl::Image2D> images{
+            cl::Image2D(context, CL_MEM_READ_WRITE, required_image_format, side,
+                        side),
+            cl::Image2D(context, CL_MEM_READ_WRITE, required_image_format, side,
+                        side),
+        };
+
+        // In each pixel of the images, the concentration of the
+        //   - U (reaction source) component is stored in the R channel,
+        //   - V (reaction result) component is stored in the G channel.
+        // The B channel is unused, and the A (alpha) channel must be set to 1,
+        // so the resulting image is visible in the image viewer.
+        //
+        // First, fill the source image with chemical U (the source component of
+        // the reaction)
+        compute_queue.enqueueFillImage(images.read,
+                                       cl_float4{ { 1.F, 0.F, 0.F, 1.F } },
+                                       { 0, 0 }, { side, side, 1 });
+
+        // In addition, fill a small rectangle in the middle of the source image
+        // with chemical V (the result of the reaction)
+        compute_queue.enqueueFillImage(
+            images.read, cl_float4{ { 1.F, 1.F, 0.F, 1.F } },
+            { side / 2, side / 2 }, { side / 100, side / 100, 1 });
+
+        // The copy_event is initialized with a completed user event,
+        // so it doesn't block the first kernel launch.
+        cl::UserEvent completed_event(context);
+        completed_event.setStatus(CL_COMPLETE);
+        cl::Event copy_event = completed_event;
+        cl::Event prev_compute_event = completed_event;
+
+        const std::size_t num_images_writes =
+            (iterations + save_at_every - 1) / save_at_every;
+        image_write_tasks.resize(num_images_writes);
+        std::intptr_t copy_job_id{};
+        for (std::size_t iter = 0; iter < iterations; ++iter)
+        {
+            // Enqueue the next step of the iteration and swap the source and
+            // destination images. This synchronizes with the previous copy of
+            // the input image, ensuring that the copy is finished before the
+            // current kernel overwrites the data.
+            auto compute_event = reaction_diffusion_step(
+                cl::EnqueueArgs(compute_queue, copy_event,
+                                cl::NDRange(side, side)),
+                images.read, images.write);
```

Comment:
This isn't something that needs to be changed in this sample, but noting for completeness: this C++ sample is making extra calls that aren't present in the C sample, mostly due to how this kernel is being enqueued:

1. Building the `EnqueueArgs` makes a copy of the queue and event, which requires retaining the queue and the passed-in event, then releasing them when the `EnqueueArgs` goes out of scope.
2. The variadic template arguments for the read and write images are passed-by-value and not passed-by-reference, so there is a retain and release for them also.

I think the fix for these would go in the C++ bindings and not in the sample.

If the C++ version of the sample is slower than the C version this is probably why.